lib/models/order_book.js

```diff
@@ -385,11 +398,11 @@ class OrderBook extends EventEmitter {
 
       if (insertIndex === -1) {
         if (
-          (dir === -1 && pl[2] < 0 && price < pl[priceI]) || // by price
-          (dir === -1 && pl[2] < 0 && price === pl[priceI] && (raw && entry[0] < pl[0])) || // by order ID
-          (dir === 1 && pl[2] > 0 && price > pl[priceI]) ||
-          (dir === 1 && pl[2] > 0 && price === pl[priceI] && (raw && entry[0] < pl[0])) ||
-          (dir === 1 && pl[2] < 0)
+          (dir === -1 && (pl.length === 4 ? -pl[3] : pl[2]) < 0 && price < pl[priceI]) || // by price
+          (dir === -1 && (pl.length === 4 ? -pl[3] : pl[2]) < 0 && price === pl[priceI] && (raw && entry[0] < pl[0])) || // by order ID
+          (dir === 1 && (pl.length === 4 ? -pl[3] : pl[2]) > 0 && price > pl[priceI]) ||
+          (dir === 1 && (pl.length === 4 ? -pl[3] : pl[2]) > 0 && price === pl[priceI] && (raw && entry[0] < pl[0])) ||
+          (dir === 1 && (pl.length === 4 ? -pl[3] : pl[2]) < 0)
```

Comment:
Why did you negate the amount here/everywhere else for funding entries?

Reply:
If it is a funding entry then bids have negative amounts and the book is sorted bids first. I see that error in checksum calculations. Is there any documentation on calculating checksums for funding books? There is an additional value of `period` in this case. Should we include it into the checksum?
I have been watching it for many hours with the option `manageOrderBooks: true`
```javascript
ws.subscribeOrderBook('fUSD', 'P0', 25);
ws.onOrderBook({}, (data) => console.log(data));
```
and on every update the book was well-sorted.

Reply:
Hmm, very strange, I'm not sure why checksums are failing then. I patched your PR locally to construct the CS string with `rate:amount` instead of `orderID:amount` for funding books, but it still fails. I noticed sometimes there are entries with a rate of `0` (asks) on `fUSD`, and they appear in the middle of the checksum string, hence the sorting idea.

There are no docs for this, but it should be `rate:amount` in the CS string

Reply:
They appear in the middle of the checksum string because of this:
https://github.com/bitfinexcom/bitfinex-api-node/blob/9cf42635de4c92bdccdd7dcbafbc3d7876748bc1/lib/models/order_book.js#L53
If current rate is low (for example  0.00000211) then in the checksum it equals 0.
So, we should change it to something like this:
```javascript
const price = raw
          ? bid[0]
          : bid.length === 4 ? Number(prepareAmount(bid[0])) : Number(preparePrice(bid[0]))
```

Reply:
I tested multiple times this with different setting and I got it valid only once 

> bfx:examples:ws2_ob_checksum recv valid cs for fUSD:R0:25 1317127459 +10s
But if I start it again then there is a mismatch.

Reply:
Could you please test the last commit. I receive `recv valid cs for fUSD:R0:25 560917535` or `recv valid cs for fUSD:R0:100 1934287952`  in 75% of messages.

Reply:
I noticed that the checksum is different only if an order with the same price is coming. The function `updateArrayOBWith` works as it is supposed to work. You can see an example [here](https://gist.github.com/vansergen/f8113efa56f76fa7753e008aede8a894). Every time I got the error `bfx:examples:ws2_ob_checksum error` there was an insertion of an order with the same price as presented in the book. For example, I received the message `bfx:examples:ws2_ob_checksum recv valid cs for fUSD:R0:25 -21786261` that the checksum is correct, then the new order `[ 325598007, 2, 0.000229, 119.50770688 ]`  was arrived. The function `updateArrayOBWith` inserted it into the book before the order `[ 325598112, 2, 0.00022899, 102.8 ]` and after the order `[ 325598032, 2, 0.000229, 211.35956946 ]`. Right after that I received the error message `bfx:examples:ws2_ob_checksum error fUSD:R0:25 OB checksum mismatch: got -56619784, want -1979834461`. The next message was to delete the order from the book `[ 325598007, 2, 0, 1 ]` and there was no error `bfx:examples:ws2_ob_checksum recv valid cs for fUSD:R0:25 1835844159`. But an order with the price presenting in the book is not always produce the checksum mismatch error.

Reply:
I've reproduced, it is correct most of the time but diverges when inserting a new order at the same rate. The insert behavior looks correct to me, I've sent a ping internally and am investigating. I'll update you when I figure it out. The checksum must be correct 100% of the time before merge, otherwise if we say that `OrderBook` supports funding symbols then users will complain/open issues due to failing checksums.

Thanks for your effort so far, this PR is great :dancer:

Reply:
@vansergen there is an issue with sorting of funding book rows on the backend, we are currently investigating. I'll merge this PR once it is fixed & checksums match, but it looks like it is correct and the problem is on the backend.